### PR TITLE
Add challenge: PPO Clipped Surrogate Loss

### PR DESCRIPTION
**Pull request by basilwong**
## Summary

  Adds **easy challenge 107: PPO Clipped Surrogate Loss**.

  The challenge implements a PPO-style token-level clipped surrogate objective with a non-negative reference-policy KL regularizer. It reduces the combined objective across a batch of fixed-length response sequences and writes one scalar loss.

  ## Scope

  - Inputs:
    - advantages, current-policy, previous-policy, and reference-policy log probabilities: `[B, S]`
  - Computes the current-to-previous policy probability ratio per token.
  - Applies the standard clipped surrogate objective using `clip_eps`.
  - Uses the non-negative `k3` reference-policy KL sample estimator:
    - it is zero when current and reference log probabilities match;
    - it penalizes divergence from the reference policy.
  - `advantages`, `log_pi_old`, and `log_ref` are fixed precomputed inputs; gradients must not flow through them.
  - Padding and token masks are outside the scope of this challenge: every token position participates in the reduction.
  - The challenge is classified as **easy** because it is a direct tensor implementation of a fully specified objective.

  ## Documentation and validation

  - Documents tensor shapes, reduction dimensions, numerical bounds, and fixed-input gradient semantics.
  - Reflows long equations and worked examples into multi-line aligned MathJax blocks for readability.
  - Adds a zero-advantage, nonzero-KL test to verify that the reference-policy penalty remains positive and is not cancelled or assigned the wrong sign.

  Modal runtime validation notebook: [PPO Clipped Surrogate Loss notebook](https://modal.com/notebooks/basilwong/main/nb-IIwN4rrliihp0j7r1xmPnX)

  ## Reference

  [Proximal Policy Optimization Algorithms — Schulman et al. (2017)](https://arxiv.org/abs/1707.06347)

### CODE REVIEW

**Comment by basilwong** [comment]
## Related Challenges

  - [Value Clipping](https://github.com/AlphaGPU/leetgpu-challenges/blob/main/challenges/easy/62_value_clipping/challenge.html)

    PPO uses clipping as one component of its policy-ratio objective. Value Clipping is an element-wise bounded-range operation, while PPO first derives a probability ratio from current and previous policy log probabilities, clips that ratio around one, combines it with an advantage, applies a reference-policy penalty, and reduces the result to a scalar loss. The shared primitive is intentional; the optimization objective is distinct.

File: challenges/easy/107_ppo_clipped_surrogate_loss/challenge.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+import math
+from typing import Any, Dict, List
+
+import torch
+from core.challenge_base import ChallengeBase, OutTensor, RandnTensor
+
+
+class Challenge(ChallengeBase):
+    name = "PPO Clipped Surrogate Loss"
+    atol = 0.001
```
Comment (shxjames) [inline challenges/easy/107_ppo_clipped_surrogate_loss/challenge.py]: This is probably too loose for tolerance; maybe change it to something tigher like 1e-5

File: challenges/easy/107_ppo_clipped_surrogate_loss/challenge.html [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+\]
+
+<h2>Constraints</h2>
+<ul>
+  <li>1 &le; <code>B</code> &le; 256</li>
+  <li>1 &le; <code>S</code> &le; 16,384</li>
+  <li>All tensors contain 32-bit floating point values</li>
+  <li><code>0 &le; clip_eps &lt; 1</code> and <code>0 &le; beta &le; 1</code></li>
+  <li>Both <code>log_pi - log_pi_old</code> and <code>log_ref - log_pi</code> lie in [-16, 16], so every exponential is finite in float32</li>
+  <li>Performance is measured with <code>B</code> = 128, <code>S</code> = 2,048</li>
```
Comment (shxjames) [inline challenges/easy/107_ppo_clipped_surrogate_loss/challenge.html]: I think this is probably insufficient for a performance test; we need this to be much bigger for performance to differ across solutions

**Comment by shxjames** [comment]
Another thing: you might want to fix the math notations; currently it's pretty messed up on the frontend, use more symbols instead
Screenshot 2026-07-27 at 16 45 35


One more suggestion: for correctness test cases, we should try to add more edge cases. For example, when we limit the clipping ratio to 0 (i.e. $$\epsilon = 0$$, even though no one in RL would actually do this)

**Comment by shxjames** [comment]
Oh and I'd classify this as medium because we put parallel scan as medium, and there's a sum here asw :)

**Comment by basilwong** [comment]
Thanks for the feedback @shxjames !!

I made the following updates in response to the review feedback:

- Cleaned up the frontend math notation and layout so the clipped-surrogate computation renders clearly in the challenge preview. It's formatted now to be easier to read as well from a flow standpoint. 

image

---

For simplicity I removed the reference-policy/KL inputs and penalty(already part of GRPO), since they are not part of the canonical clipped-surrogate objective this challenge is intended to teach. 

1. Aligned the challenge with canonical PPO’s clipped-surrogate objective:

    `-mean(min(ratio * advantage, clip(ratio, 1 - ε, 1 + ε) * advantage))`

2. Updated the reference implementations, starter templates, and functional-test interface to use only current and old policy log-probabilities, advantages, and `clip_eps`.

---

Added functional edge cases for `clip_eps = 0`, clipping-inactive positive/negative advantage directions, and nonzero multi-batch reduction.

---

Updated the [Modal notebook](https://modal.com/notebooks/basilwong/main/nb-IIwN4rrliihp0j7r1xmPnX) to deterministically verify that the example implementations pass the functional tests.

**Review by shxjames — APPROVED** [review]
